mkdir_linux.go

```diff
@@ -0,0 +1,224 @@
+			// hit a symlink or some other non-directory component from the
+			// error, but we have to readlink the target anyway so we can check
+			// that way so assume it's a symlink.
+			linkDest, err := readlinkatFile(currentDir, part)
+			if err != nil {
+				if errors.Is(err, unix.EINVAL) {
+					// The part was not a symlink, so assume that it's a
+					// regular file. It is possible for it to be a directory
+					// (if an attacker is swapping a directory and
+					// non-directory at this subpath) but erroring out here is
+					// better anyway.
+					err = fmt.Errorf("path component is invalid: %w", unix.ENOTDIR)
+				}
+				return nil, "", err
+			}
+
+			linksWalked++
+			if linksWalked > maxSymlinkLimit {
+				return nil, "", &os.PathError{Op: "partialOpen", Path: root + "/" + unsafePath, Err: unix.ELOOP}
+			}
+
+			remainingPath = linkDest + "/" + remainingPath
+			// Absolute symlinks reset any work we've already done.
+			if path.IsAbs(linkDest) {
+				nextDir, err := dupFile(rootDir)
+				if err != nil {
+					return nil, "", fmt.Errorf("clone root fd: %w", err)
+				}
+				_ = currentDir.Close()
+				currentDir = nextDir
+				currentPath = "/"
+			}
+
+		default:
+			return nil, "", err
+		}
+	}
+	// All of the components existed!
+	return currentDir, "", nil
+}
+
+// MkdirAll is a race-safe alternative to the Go stdlib's os.MkdirAll function,
+// where the new directory is guaranteed to be within the root directory (if an
+// attacker can move directories from inside the root to outside the root, the
+// created directory tree might be outside of the root but the key constraint
+// is that at no point will we walk outside of the directory tree we are
+// creating).
+//
+// Effectively, MkdirAll(root, unsafePath, mode) is equivalent to
+//
+//	path, _ := securejoin.SecureJoin(root, unsafePath)
+//	err := os.MkdirAll(path, mode)
+//
+// But is much safer. The above implementation is unsafe because if an attacker
+// can modify the filesystem tree between SecureJoin and MkdirAll, it is
+// possible for MkdirAll to resolve unsafe symlink components and create
+// directories outside of the root.
+func MkdirAll(root, unsafePath string, mode os.FileMode) error {
+	// Try to open as much of the path as possible.
+	currentDir, remainingPath, err := partialOpen(root, unsafePath)
+	if err != nil {
+		return fmt.Errorf("find existing subpath: %w", err)
+	}
+	defer currentDir.Close()
+
+	// The remaining path can be cleaned because none of the path components
+	// exist and thus there are no symlinks to worry about (and removing ..
+	// also removes the possibility of a parallel-write attack tricking us).
+	remainingPath = path.Join("/", filepath.ToSlash(remainingPath))
+
+	// Create the remaining components.
+	for _, part := range strings.Split(remainingPath, "/") {
+		switch part {
+		case "", ".":
+			// Skip no-op components. There should be none, but check anyway.
+			continue
```

Comment:
nit: error-out here too seems safer. We really don't expect this.

Reply:
Fair enough.